src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -1697,6 +1742,17 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
     }
 
+    uint64_t ControlCore::SwapChainHandle() const
+    {
+        // This is called by:
+        // * TermControl::RenderEngineSwapChainChanged, who is only registered
+        //   after Core::Initialize() is called.
+        // * TermControl::_InitializeTerminal, after the call to Initialize, for
+        //   _AttachDxgiSwapChainToXaml.
+        // In both cases, we'll have a _renderEngine by then.
+        return _renderEngine ? reinterpret_cast<uint64_t>(_renderEngine->GetSwapChainHandle()) : 0u;

```

Comment:
This code is unsafe, because `GetSwapChainHandle` returns a weak reference as a `HANDLE`, while the renderer thread might destroy and recreate the swap chain at any random time (this is true for both AtlasEngine and DxRenderer). WinRT lacks a safe way of transmitting handles (COM got it 🥲), so I think the next best thing is to `AddRef` the handle (or `.detach()` the com ptr) transmit it to the other side / the caller, who then adopts it and `Release`s it when it's done using it (or `.attach()`). This should work the same with both `wil::unique_handle` and `winrt::handle` (which both got `attach` / `detach`).

Also, I think you need to hold the console lock here, because reading the HANDLE might otherwise read stale data (only reproducible on ARM, because on x64 all pointer-sized reads are always atomic).

I haven't looked at this PR in full, but to be honest, I'd personally prefer if this function didn't exist at all to begin with, since I'm planning to reduce the amount of locking/mutexes and interactions between the UI code and the renderer. We already have the `SetCallback` function which results in the calls to `TermControl::RenderEngineSwapChainChanged`. There, once it has resumed on the UI thread, we could store the last `HANDLE` we've received from the renderer as a member on `TermControl` and instead of calling `term->_core.SwapChainHandle()` you'd call `term->SwapChainHandle()` which then returns that cached handler. Since `SwapChainHandle()` would presumably be called on the UI thread as well, it wouldn't even require any locking.